### PR TITLE
Add TORCH_ASSERT_ONLY_METHOD_OPERATORS macro

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69840
* #68462
* #68693
* #68692
* #68691
* #68697
* #68690
* #68689
* #68714
* __->__ #68688
* #68687
* #68247

This adds a new macro `TORCH_ASSERT_ONLY_METHOD_OPERATORS` which
allows `Tensor.h` to be included, but not headers which pull in all
other operators. So, a file that defines this macro needs to use the
fine-grained headers to include only the operators being used.

Differential Revision: [D32596267](https://our.internmc.facebook.com/intern/diff/D32596267)

### CODE REVIEW

File: aten/src/ATen/templates/DispatchKeyFunctions_inl.h [hunk 1]
```diff
@@ -7,6 +7,14 @@
 #include <c10/core/Scalar.h>
 #include <ATen/core/Reduction.h>
 
+// TODO: If necessary, consider adding <ATen/ops/{function}_key.h> headers
```
Comment (albanD) [inline aten/src/ATen/templates/DispatchKeyFunctions_inl.h:10]: Is this for you or future devs?

Do we want to write a "note" somewhere about the function specific headers and have all the error message point to it?

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596267).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596267).

File: aten/src/ATen/templates/DispatchKeyFunctions_inl.h (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/templates/DispatchKeyFunctions_inl.h:10]: Sorry, just noticed my comment didn't get posted. This was a note to myself, referring to splitting this header into per-operator headers. The next PR in the stack resolves that and the comment is gone.

**Review by albanD — APPROVED** [review]

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596267).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596267).

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**